### PR TITLE
test: standardized unit testing with jest

### PR DESCRIPTION
**Pull request by NathanWalker**
## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/tools/notes/CONTRIBUTING.md#commit-messages.
- [ ] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [x] You have signed the [CLA](http://www.nativescript.org/cla).
- [x] All existing tests are passing: https://github.com/NativeScript/NativeScript/blob/master/tools/notes/DevelopmentWorkflow.md#running-unit-tests-application.
- [x] Tests for the changes are included - https://github.com/NativeScript/NativeScript/blob/master/tools/notes/WritingUnitTests.md.

## What is the current behavior?

Unit tests were awkward to work with and maintain.

## What is the new behavior?

Unit tests are now easy to work with, use and maintain.

### CODE REVIEW

File: packages/core/jest.config.ts [hunk 1]
```diff
@@ -1,5 +1,27 @@
...
+			'ts-jest',
+			{
+				useESM: true,
+				tsconfig: '<rootDir>/tsconfig.spec.json',
+			},
+		],
+	},
+	extensionsToTreatAsEsm: ['.ts'],
+	moduleDirectories: ['node_modules'],
+	moduleFileExtensions: ['ts', 'js', 'ios.ts'],
```
Comment (rigor789) [inline packages/core/jest.config.ts:21]: Wonder if we could run the whole test-suite twice, once against `.ios.ts` and once against `.android.ts` to cover all files...

File: packages/core/global-types.d.ts [hunk 2]
```diff
@@ -350,13 +350,6 @@ declare function fail(data: any): void;
 
 declare function zonedCallback(callback: Function): Function;
 
-declare class WeakRef<T> {
-	constructor(obj: T);
-	get(): T;
-	clear(): void;
-	deref(): T | undefined;
-}
```
Comment (rigor789) [inline packages/core/global-types.d.ts:358]: Should we keep this, and mark `clear` as `@deprecated`?

**Review by rigor789 — COMMENTED** [review]
Looks great, noted a few questions here&there...

File: packages/core/jest.config.ts (hunk 1 above)
Comment (NathanWalker) [reply packages/core/jest.config.ts:21]: Thanks for mentioning yes I am wanting to do this. Was considering two different configs and .spec.json that could be run independently for this.

File: packages/core/global-types.d.ts (hunk 2 above)
Comment (NathanWalker) [reply packages/core/global-types.d.ts:358]: Good idea yes will add this back. I initially removed to see all spots in core that used .get - Android is not operational at moment the way the changes are here and was going to try a polyfill to keep the .get behavior but not sure which is best yet.

Comment (NathanWalker) [reply packages/core/global-types.d.ts:358]: added back with proper deprecation note here: https://github.com/NativeScript/NativeScript/pull/10047/commits/05651b1ec330b049b240e9745eab3174ef8bd198

File: packages/core/color/color-common.ts [hunk 3]
```diff
@@ -177,7 +177,7 @@ export class Color implements definition.Color {
 		return this.hex;
 	}
 
-	public static fromIosColor(value: UIColor): Color {
+	public static fromIosColor(value: any /* UIColor */): Color {
```
[the remark this answers is not in the document]
Comment (NathanWalker) [reply packages/core/color/color-common.ts]: We'll try it, added it here: https://github.com/NativeScript/NativeScript/pull/10047/commits/05651b1ec330b049b240e9745eab3174ef8bd198

File: package.json [hunk 4]
```diff
@@ -61,9 +61,10 @@
     "sass": "^1.45.2",
     "shady-css-parser": "^0.1.0",
     "tree-kill": "^1.2.2",
-    "ts-jest": "28.0.8",
+    "ts-jest": "29.0.3",
     "ts-node": "10.9.1",
     "ts-patch": "^2.0.1",
+    "tslib": "^2.3.0",
```
[the remark this answers is not in the document]
Comment (NathanWalker) [reply package.json]: Thanks for note, cleaned up here: https://github.com/NativeScript/NativeScript/pull/10047/commits/05651b1ec330b049b240e9745eab3174ef8bd198

File: packages/core/global-types.d.ts (hunk 2 above)
Comment (rigor789) [reply packages/core/global-types.d.ts:358]: We also need to add a deprecation note to `clear`

Comment (NathanWalker) [reply packages/core/global-types.d.ts:358]: added here: 2154b710c102688a2ab61979dd15b167d1d277ff

File: packages/core/css/parser.spec.ts [hunk 5]
```diff
@@ -42,13 +43,13 @@ describe('css', () => {
...
+				test(parseColor, '  #45678985 ', { start: 0, end: 12, value: new Color(0x85456789) });
 				test(parseColor, '  rgb(255, 8, 128) ', { start: 0, end: 18, value: new Color(0xffff0880) });
-				test(parseColor, '  rgb(255 8 128 / .5) ', { start: 0, end: 22, value: new Color(0x80ff0880) });
+				test(parseColor, '  rgb(255 8 128 / .5) ', { start: 0, end: 21, value: new Color(0x80ff0880) });
 				test(parseColor, '  rgb(255 8 128/0.5)', { start: 0, end: 20, value: new Color(0x80ff0880) });
 				test(parseColor, '  rgba(255, 8, 128, 0.5) ', { start: 0, end: 24, value: new Color(0x80ff0880) });
-				test(parseColor, '  hsl(330.9, 100%, 51.6%) ', { start: 0, end: 25, value: new Color(0xffff0880) });
-				test(parseColor, '  hsla(330.9, 100%, 51.6%, 0.5) ', { start: 0, end: 31, value: new Color(0x80ff0880) });
+				test(parseColor, '  hsl(330.9, 100%, 51.6%) ', { start: 0, end: 25, value: new Color('hsl(330.9, 100%, 51.6%)') });
+				test(parseColor, '  hsla(330.9, 100%, 51.6%, 0.5) ', { start: 0, end: 31, value: new Color('hsla(330.9, 100%, 51.6%, 0.5)') });
```
[the remark this answers is not in the document]
Comment (NathanWalker) [reply packages/core/css/parser.spec.ts]: Fixed now.

File: packages/core/css/parser.spec.ts [hunk 6]
```diff
@@ -141,13 +142,13 @@ describe('css', () => {
 						},
 					},
 				});
-				test(parseBackground, 'linear-gradient(0deg, blue, green 40%, red)', {
+				test(parseBackground, 'linear-gradient(0deg, #333, #999 40%, #888)', {
```
[the remark this answers is not in the document]
Comment (NathanWalker) [reply packages/core/css/parser.spec.ts]: Fixed.

**Pull request merged**